larpandoracontent/LArTwoDReco/LArCosmicRay/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,877 @@
+    const ClusterList originalClusterList(1, pCluster);
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::InitializeFragmentation(*this, originalClusterList, originalListName, fragmentListName));
+
+    const Cluster *pMainTrackCluster(nullptr), *pAboveCluster(nullptr), *pBelowCluster(nullptr);
+
+    const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+    for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
+    {
+        for (const CaloHit *const pCaloHit : *mapEntry.second)
+        {
+            const CartesianVector &hitPosition(pCaloHit->GetPositionVector());
+            float thisL(0.f), thisT(0.f);
+
+            microFitResult.GetLocalPosition(pCaloHit->GetPositionVector(), thisL, thisT);
+
+            const bool isAnExtrapolatedHit(std::find(extrapolatedCaloHitIter->second.begin(), extrapolatedCaloHitIter->second.end(), pCaloHit) != extrapolatedCaloHitIter->second.end());
+            const bool isAbove(((clusterGradient * hitPosition.GetX()) + clusterIntercept) < (isVertical ? hitPosition.GetX() : hitPosition.GetZ()));
+            const bool isToRemove(!isAnExtrapolatedHit && (((thisL > rL) && isUpstream) || ((thisL < rL) && !isUpstream)));
+
+            const Cluster *&pClusterToModify(isToRemove ? (isAbove ? pAboveCluster : pBelowCluster) : pMainTrackCluster);
+
+            if (pClusterToModify)
+            {
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::AddToCluster(*this, pClusterToModify, pCaloHit));
+            }
+            else
+            {
+                PandoraContentApi::Cluster::Parameters parameters;
+                parameters.m_caloHitList.push_back(pCaloHit);
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Cluster::Create(*this, parameters, pClusterToModify));
+
+                if (pClusterToModify != pMainTrackCluster)
+                    remnantClusterList.push_back(pClusterToModify);
+            }
+        }
+    }
+
+    this->UpdateForClusterDeletion(pCluster, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
+
+    // End fragmentation
+    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, fragmentListName, originalListName));
+    return pMainTrackCluster;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::AddHitsToMainTrack(const Cluster *const pShowerCluster, const Cluster *const pMainTrackCluster, const CaloHitList &caloHitsToMerge,
+    const ClusterAssociation &clusterAssociation, ClusterList &remnantClusterList) const
+{
+    // To ignore crossing CR muon or test beam tracks
+    if ((static_cast<float>(caloHitsToMerge.size()) / static_cast<float>(pShowerCluster->GetNCaloHits())) < m_minHitFractionForHitRemoval)
```

Comment:
Any chance can have a shower cluster with no hits incoming? (Can only happen if they cluster has some isolated hits, but no non-isolated hits).

Reply:
I'm pretty sure the answer here is no. The pShowerCluster is always a cluster that has a hit identified as being along the line that connects the split clusters. This tagging is made by looking at the cluster's ordered calo hit list and making a position check on each calo hit. So, as long as the ordered calo hit list (from pCluster->GetOrderedCaloHitList) only contains non-isolated hits then we are good. I think this is true since you mentioned that isolated calo hits do not contribute to the spatial info of the cluster